[SPARK-2669] [yarn] Distribute client configuration to AM.

vanzin opened the pull request:
Currently, when Spark launches the Yarn AM, the process will use
the local Hadoop configuration on the node where the AM launches,
if one is present. A more correct approach is to use the same
configuration used to launch the Spark job, since the user may
have made modifications (such as adding app-specific configs).

The approach taken here is to use the distributed cache to make
all files in the Hadoop configuration directory available to the
AM. This is a little overkill since only the AM needs them (the
executors use the broadcast Hadoop configuration from the driver),
but is the easier approach.

Even though only a few files in that directory may end up being
used, all of them are uploaded. This allows supporting use cases
such as when auxiliary configuration files are used for SSL
configuration, or when uploading a Hive configuration directory.
Not all of these may be reflected in a o.a.h.conf.Configuration object,
but may be needed when a driver in cluster mode instantiates, for
example, a HiveConf object instead.

Commit:
[SPARK-2669] [yarn] Distribute client configuration to AM.

Currently, when Spark launches the Yarn AM, the process will use
the local Hadoop configuration on the node where the AM launches,
if one is present. A more correct approach is to use the same
configuration used to launch the Spark job, since the user may
have made modifications (such as adding app-specific configs).

The approach taken here is to use the distributed cache to make
all files in the Hadoop configuration directory available to the
AM. This is a little overkill since only the AM needs them (the
executors use the broadcast Hadoop configuration from the driver),
but is the easier approach.

Even though only a few files in that directory may end up being
used, all of them are uploaded. This allows supporting use cases
such as when auxiliary configuration files are used for SSL
configuration, or when uploading a Hive configuration directory.
Not all of these may be reflected in a o.a.h.conf.Configuration object,
but may be needed when a driver in cluster mode instantiates, for
example, a HiveConf object instead.

vanzin commented:
Note that the approach taken here is a little different than the one used in #1574 (which was seemingly abandoned).

This is the code I used for testing on a live cluster:
https://gist.github.com/vanzin/25dc9fcf14352dbb0c91

I tested both by modifying the local Hadoop config files and by setting `spark.hadoop.blah` on the command line.

SparkQA commented:
[Test build #25902 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25902/consoleFull) for   PR 4142 at commit [`79221c7`](https://github.com/apache/spark/commit/79221c712e9d8794a3e36f08fcd29f247bb9138f).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25902 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25902/consoleFull) for   PR 4142 at commit [`79221c7`](https://github.com/apache/spark/commit/79221c712e9d8794a3e36f08fcd29f247bb9138f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25902/
Test FAILed.

vanzin commented:
Can't find any error in that test output? Let's try again...

vanzin commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #25912 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25912/consoleFull) for   PR 4142 at commit [`79221c7`](https://github.com/apache/spark/commit/79221c712e9d8794a3e36f08fcd29f247bb9138f).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25912 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25912/consoleFull) for   PR 4142 at commit [`79221c7`](https://github.com/apache/spark/commit/79221c712e9d8794a3e36f08fcd29f247bb9138f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25912/
Test PASSed.

redbaron commented:
IMHO Spark should mimic what Hadoop MR is doing, which is building a Configuration object and then serializing into xml file and then shipping this xml to workers. It allows to make executors config to be entirely client-driven.

Copying files verbatim from the directory leads to not being able to override hadoop config params via "-D<param>=<value>" switch (don't know if it is supported by spark, but if it is not, it should be) or can cause surprises if config is influenced by other means (for instance different defaults for client side version of libs).

Both original patch and this one suffer from more serious problem (not saying that it should prevent this patch from being merged or anything like that, just sharing my experience): different parts of spark may request different configs to be shipped to AM, like Spark SQL requires hive-site.xml in a similar way. 

I am not familiar with spark internals, but if SparkContext instance is available to submitting client, then it can "query" it to get list of configs to push to AM, then users of SparkContext (like SQLContext) can register additional configs to be transferred at job deploy time.

vanzin commented:
Hi @redbaron ,

> IMHO Spark should mimic what Hadoop MR is doing, which is building a Configuration object and then serializing into xml

That has the limitation of not loading config data from files that the Configuration object you chose do not load. For example, if you're going to run a job that uses a HiveContext, you might want to also use hive-site.xml, as yourself mentioned in your comment. My patch covers that, since if you point `HADOOP_CONF_DIR` at a Hive config directory, it will ship that file and all should work.

> Copying files verbatim from the directory leads to not being able to override hadoop config params via "-D=" switch

If you're talking about the `-D` switch for the JVM, then that's already covered. `--driver-java-options` is reflected in the AM when running in cluster mode. There's a similar option for client mode, in case you also need that for the client-mode AM.

Feels to me like my approach covers your concerns.

Commit:
Merge branch 'master' into SPARK-2669

Conflicts:
	yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala
	yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala

SparkQA commented:
[Test build #26942 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26942/consoleFull) for   PR 4142 at commit [`a77ddd5`](https://github.com/apache/spark/commit/a77ddd51234ebf33c0bbd690ce15939dc2bc8e90).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26942 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26942/consoleFull) for   PR 4142 at commit [`a77ddd5`](https://github.com/apache/spark/commit/a77ddd51234ebf33c0bbd690ce15939dc2bc8e90).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RegisterExecutor(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26942/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-2669

Conflicts:
	yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala
	yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala

SparkQA commented:
[Test build #27310 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27310/consoleFull) for   PR 4142 at commit [`022a688`](https://github.com/apache/spark/commit/022a6884eb0422d66308484294f8b988c5a84a11).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27310 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27310/consoleFull) for   PR 4142 at commit [`022a688`](https://github.com/apache/spark/commit/022a6884eb0422d66308484294f8b988c5a84a11).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27310/
Test FAILed.

vanzin commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #27313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27313/consoleFull) for   PR 4142 at commit [`022a688`](https://github.com/apache/spark/commit/022a6884eb0422d66308484294f8b988c5a84a11).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27313 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27313/consoleFull) for   PR 4142 at commit [`022a688`](https://github.com/apache/spark/commit/022a6884eb0422d66308484294f8b988c5a84a11).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27313/
Test FAILed.

Commit:
Fix test.

The test was setting HADOOP_CONF_DIR to the temp directory, causing
the same file to be distributed via Yarn's cache multiple times. Yarn
doesn't seem to like that; so fix the test and add a check in Client.scala
to detect that situation and fail early when it happens.

SparkQA commented:
[Test build #27650 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27650/consoleFull) for   PR 4142 at commit [`34bdbd8`](https://github.com/apache/spark/commit/34bdbd81ca95fc4b844686f852a06b2c937d522c).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27650 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27650/consoleFull) for   PR 4142 at commit [`34bdbd8`](https://github.com/apache/spark/commit/34bdbd81ca95fc4b844686f852a06b2c937d522c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27650/
Test PASSed.

vanzin commented:
Ping.

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -233,6 +229,15 @@ private[spark] class Client(
+    def addDistributedUri(uri: URI): Unit = {
+      val uriStr = uri.toString()
+      if (distributedUris.contains(uriStr)) {
+        throw new IllegalArgumentException(
```
Do we need to throw an exception here?  What's the harm in just logging a warning and not adding the URI?

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -216,6 +211,7 @@ private[spark] class Client(
     val fs = FileSystem.get(hadoopConf)
     val dst = new Path(fs.getHomeDirectory(), appStagingDir)
     val nns = getNameNodesToAccess(sparkConf) + dst
+    val distributedUris = new HashSet[String]
```
The use of this isn't immediately clear.  Mind adding a short comment above describing its purpose?

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -263,6 +269,45 @@ private[spark] class Client(
+    hadoopConfFiles.foreach { case (name, file) =>
+      addDistributedUri(file.toURI())
+      val destPath = copyFileToRemote(hadoopConfPath, new Path(file.toURI()), replication)
+      distCacheMgr.addResource(fs, hadoopConf, destPath,
```
Arguments should each have their own line.

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -679,6 +725,9 @@ object Client extends Logging {
   // Distribution-defined classpath to add to processes
   val ENV_DIST_CLASSPATH = "SPARK_DIST_CLASSPATH"
 
+  // Subdirectory where the user's hadoop config files are written in the job staging dir.
```
"job staging dir" -> "app staging dir"

sryza commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala:
```diff
@@ -120,6 +121,8 @@ class YarnClusterSuite extends FunSuite with BeforeAndAfterAll with Matchers wit
 
     fakeSparkJar = File.createTempFile("sparkJar", null, tempDir)
+    hadoopConfDir = new File(tempDir, Client.HADOOP_CONF_DIR)
+    assert(hadoopConfDir.mkdir())
```
Can we clean this dir up after the test?

vanzin replied:
The directory is a temp dir (so unique to this suite) and cleaned up when the JVM shuts down (through hooks set up by the Utils class). Is there anything else you were thinking of here?

sryza replied:
Nope, that covers it.

sryza commented:
This change seems reasonable to me. I had an offline conversation with Marcelo about edge cases and there was nothing I could think of.

The biggest concern here is that we now require the edge node that the client is submitting their app from to have the configuration files that the app should run with.  This is to some extent required already - the edge node must already have the Hadoop configs for the ResourceManager address and NameNode address so that it can submit apps to the cluster and add files to the distributed cache.  So I think this is a reasonable expectation.  However, it would good to call it out explicitly in the Spark on YARN doc.

Commit:
Review feedback.

Commit:
Merge branch 'master' into SPARK-2669

Commit:
Remove stale test.

SparkQA commented:
[Test build #28803 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28803/consoleFull) for   PR 4142 at commit [`cbb9fb3`](https://github.com/apache/spark/commit/cbb9fb3b16f3711cae8d3eae8c7573a88d294798).
 * This patch merges cleanly.

tgravescs commented:
I think this is a good change and let us set some hadoop configs easier, but I'm a bit concerned about the extra load this could cause if you have a lot of files in the conf dir.  By load I mean on the namenode as well as just task startup time.  The files should all be relatively small so I wouldn't think the startup time would be affected to much, but I'm curious if you measured it at all?  Some hadoop config dirs could potentially have a lot of files in them.  There are metrics property files, the normal xml files, env files, and whatever else happens to be thrown in there. For example I looked at one cluster and saw 41 files in that directory.   This could cause a bunch of unneeded load on the namenode, especially if you are starting thousands of executors all at once.  MR obviously packages the confs and send them but its just in one job.xml file.  

If these really aren't needed on the executors then perhaps we should just distribute to the AM. This is possible you probably just need to add another env/config. We used to do this for certain files when we used sparks distribution method for jars.

Even though I don't like adding more configs I'm thinking it would be nice to have switch to turn this off and it to use the ones on the cluster like it did before.

SparkQA commented:
[Test build #28803 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28803/consoleFull) for   PR 4142 at commit [`cbb9fb3`](https://github.com/apache/spark/commit/cbb9fb3b16f3711cae8d3eae8c7573a88d294798).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class JavaDoubleRDD(val srdd: RDD[scala.Double])`
  * `class KMeansModel(Saveable, Loader):`
  * `                logError("User class threw exception: " + cause, cause)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28803/
Test PASSed.

vanzin commented:
Our config directories don't generally contain a lot of files (a handful usually), but I can see it being different for different people.

My initial gut feeling is that at least the HDFS config files, and perhaps the Yarn ones, are needed on the executors. Others (like hive-site.xml) are only needed by the driver (not even the AM). Let me see if there's a clean way to separate those out - then this change could be a hybrid of the original patch, which basically serialized the YarnConfiguration object, and this one for the other conf files.

Commit:
Merge branch 'master' into SPARK-2669

Commit:
Zip all config files and upload them as an archive.

This avoids creating lots of small files in the NN when
starting up the job. The archive uses the lowest compression
level so that it's faster.

vanzin commented:
See updated patch; the code now uploads a single file as `LocalResourceType.ARCHIVE`, so that the containers see the contents as files (instead of a zip file in the classpath). This should avoid having conf directories with lots of files causing undue burden on the NN.

SparkQA commented:
[Test build #29628 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29628/consoleFull) for   PR 4142 at commit [`ed45b7d`](https://github.com/apache/spark/commit/ed45b7df3e636794fae139ebecef81c3da949aeb).

SparkQA commented:
[Test build #29628 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29628/consoleFull) for   PR 4142 at commit [`ed45b7d`](https://github.com/apache/spark/commit/ed45b7df3e636794fae139ebecef81c3da949aeb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29628/
Test FAILed.

vanzin commented:
This test has been a little flaky lately, need to look at it...

vanzin commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #29631 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29631/consoleFull) for   PR 4142 at commit [`ed45b7d`](https://github.com/apache/spark/commit/ed45b7df3e636794fae139ebecef81c3da949aeb).

SparkQA commented:
[Test build #29631 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29631/consoleFull) for   PR 4142 at commit [`ed45b7d`](https://github.com/apache/spark/commit/ed45b7df3e636794fae139ebecef81c3da949aeb).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29631/
Test PASSed.

sryza commented:
Is my understanding correct that the only change since the last version is that we bundle all the files together instead of uploading them individually?

Can we break up the `prepareLocalResources` method?  It's looking huge.

vanzin commented:
Yes, that's the only change.

I'd rather not do any refactoring in this change. Client.scala needs a thorough cleanup, it's better to do that as a separate task that looks at the code as a whole.

sryza commented:
I don't think the whole method needs to be refactored right now, but there's a 50-line contiguous addition to it.  Can that go in its own method?

vanzin commented:
I'm not saying the whole method needs refactoring, I'm saying the whole file needs refactoring. Putting that block - which, semantically, does really belong in that method - in a separate method won't achieve anything to clean things up.

sryza commented:
I'm advocating for a sub-method `private def createHadoopConfArchive(): File` to house collecting and zipping the conf files.

This stuff obviously makes sense underneath `prepareLocalResources`, but doesn't need to go in its main-line.

Commit:
Le sigh.

SparkQA commented:
[Test build #29766 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29766/consoleFull) for   PR 4142 at commit [`f693152`](https://github.com/apache/spark/commit/f6931521b239ba89cb961ddd1f3c529681568f42).

SparkQA commented:
[Test build #29766 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29766/consoleFull) for   PR 4142 at commit [`f693152`](https://github.com/apache/spark/commit/f6931521b239ba89cb961ddd1f3c529681568f42).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29766/
Test PASSed.

sryza commented:
LGTM

vanzin commented:
@tgravescs any comments on the latest patch?

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -303,6 +326,59 @@ private[spark] class Client(
+    }
+
+    if (!hadoopConfFiles.isEmpty) {
+      val hadoopConfArchive = File.createTempFile(HADOOP_CONF_DIR, ".jar",
```
any reason we use .jar instead of .zip?

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -303,6 +326,59 @@ private[spark] class Client(
+   * Create an archive with the Hadoop config files for distribution.
+   *
+   * These are only really used by the AM, since executors will use the configuration object
+   * broadcast by the driver. But this is the easiest way to make sure the files are available for
```
So is this comment true, is it only really needed by the AM or do we need it on the executors for things like hive-site.xml?  If its true then we should change our call to ClientDistributedCacheManager.addResource to pass in the flag to make it appMasterOnly.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -679,6 +755,9 @@ object Client extends Logging {
   val ENV_DIST_CLASSPATH = "SPARK_DIST_CLASSPATH"
 
+  // Subdirectory where the user's hadoop config files are written in the app staging dir.
+  val HADOOP_CONF_DIR = "__hadoop_conf__"
```
HADOOP_CONF_DIR is the name hadoop uses for the env variable.  To avoid confusion with that could we rename this variable?

Also (minor nit) you might want to remove the "in the app staging dir" part of the comment.  Perhaps replace with when localized.  When its in the app staging dir its call __hadoop_conf__.zip, then it gets localized then it would go into the directory.

Commit:
Review feedback.

Commit:
Merge branch 'master' into SPARK-2669

Conflicts:
	yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

SparkQA commented:
[Test build #30197 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30197/consoleFull) for   PR 4142 at commit [`f5434b9`](https://github.com/apache/spark/commit/f5434b9c31f416cff515aee9cb5a71e80aedb405).

SparkQA commented:
[Test build #30197 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30197/consoleFull) for   PR 4142 at commit [`f5434b9`](https://github.com/apache/spark/commit/f5434b9c31f416cff515aee9cb5a71e80aedb405).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30197/
Test PASSed.

vanzin commented:
Hi @tgravescs , how's the latest patch? Thanks

tgravescs commented:
Changes look good. Thanks @vanzin

litao-buptsse commented:
@vanzin I think it will be nice to allow user to choose use hadoop configuration on cluster-side or on client-side. In my company's hadoop cluster, when I use this feature in spark 1.4, it will trigger some bugs to use hadoop configuration on client-side (it's our hadoop bugs, not spark). So if I can choose whether use this feature or not would be nice to me.
